homeassistant/components/crownstone/__init__.py

```diff
@@ -2,25 +2,42 @@
 
 from __future__ import annotations
 
-from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
 from homeassistant.core import HomeAssistant
 
-from .const import DOMAIN
-from .entry_manager import CrownstoneEntryManager
+from .const import PLATFORMS
+from .entry_manager import CrownstoneConfigEntry, CrownstoneEntryManager
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_setup_entry(hass: HomeAssistant, entry: CrownstoneConfigEntry) -> bool:
     """Initiate setup for a Crownstone config entry."""
     manager = CrownstoneEntryManager(hass, entry)
 
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = manager
+    if not await manager.async_setup():
+        return False
 
-    return await manager.async_setup()
+    entry.runtime_data = manager
 
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    # HA specific listeners
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, manager.on_shutdown)
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: CrownstoneConfigEntry) -> bool:
     """Unload a config entry."""
-    unload_ok: bool = await hass.data[DOMAIN][entry.entry_id].async_unload()
-    if len(hass.data[DOMAIN]) == 0:
-        hass.data.pop(DOMAIN)
-    return unload_ok
+    entry.runtime_data.async_unload()
+
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def _async_update_listener(
```

Comment:
```suggestion
async def async_update_listener(
```

Not really necessary to be private.

_Related: Should the name be standardized similar to `async_setup_entry` and `async_unload_entry`?That could make the `entry.async_on_unload` call unnecessary._